File: src/waypalm-site/src/main/java/waypalm/site/web/mvc/oauth/OAuthAuthenticationFilter.java
package waypalm.site.web.mvc.oauth;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.core.Authentication;
import org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter;
import org.springframework.security.web.authentication.rememberme.AbstractRememberMeServices;
import org.springframework.social.connect.Connection;
import org.springframework.social.connect.ConnectionFactory;
import org.springframework.social.connect.ConnectionFactoryLocator;
import org.springframework.social.connect.support.OAuth1ConnectionFactory;
import org.springframework.social.connect.support.OAuth2ConnectionFactory;
import org.springframework.social.support.URIBuilder;
import org.springframework.util.Assert;
import org.springframework.util.StringUtils;

import javax.inject.Inject;
import javax.security.sasl.AuthenticationException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.http.HttpSession;
import java.io.IOException;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

public class OAuthAuthenticationFilter extends AbstractAuthenticationProcessingFilter {
    private static final String FILTER_ROUTE = "/j_spring_oauth_security_check";
    private static final String OAUTH_PARAM_PROVIDER = "oauth_provider";

    private Set<String> returnToUrlParameters = Collections.emptySet();

    public OAuthAuthenticationFilter() {
        super(FILTER_ROUTE);
    }

    @Inject
    ConnectionFactoryLocator connectionFactoryLocator;
    @Value("${setup.social.connect.overrideSecureUrl}")
    String overrideSecureUrl;

    public void setReturnToUrlParameters(Set<String> returnToUrlParameters) {
        Assert.notNull(returnToUrlParameters, "returnToUrlParameters cannot be null");
        this.returnToUrlParameters = returnToUrlParameters;
    }

    @Override
    public void afterPropertiesSet() {
        super.afterPropertiesSet();
        if (returnToUrlParameters.isEmpty() && getRememberMeServices() instanceof AbstractRememberMeServices) {
            returnToUrlParameters = new HashSet<>();
            returnToUrlParameters.add(((AbstractRememberMeServices) getRememberMeServices()).getParameter());
        }
    }

    /**
     * OAuth Authentication flow: See http://dev.twitter.com/pages/auth
     */
    @Override
    public Authentication attemptAuthentication(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String providerName = request.getParameter(OAUTH_PARAM_PROVIDER);
        if (providerName == null) {
            providerName = retrieveProviderName(request);
        }

        if (providerName == null) {
            throw new IllegalStateException("OAuth provider is unknown");
        }

        OAuthFlow flow = selectFlow(providerName, request);
        if (flow == null) {
            throw new AuthenticationException("unknown OAuth flow");
        }

        flow.setCallbackUrl(callbackUrl(request));

        if (flow.isInitiationRequired()) {
            storeProviderName(request, providerName);
            flow.initiateAuthentication(request, response);
            return null;
        }

        Connection<?> connection = flow.attemptAuthentication(request);
        OAuthAuthenticationToken token = new OAuthAuthenticationToken(connection);
        return getAuthenticationManager().authenticate(token);
    }

    private OAuthFlow selectFlow(String provider, HttpServletRequest request) throws IOException {
        ConnectionFactory<?> connectionFactory = connectionFactoryLocator.getConnectionFactory(provider);

        if (StringUtils.hasText(request.getParameter("oauth_token"))) {
            return new OAuth1Flow(false, (OAuth1ConnectionFactory<?>) connectionFactory);
        } else if (StringUtils.hasText(request.getParameter("code"))) {
            return new OAuth2Flow(false, (OAuth2ConnectionFactory<?>) connectionFactory);
        } else if (connectionFactory instanceof OAuth2ConnectionFactory) {
            return new OAuth2Flow(true, (OAuth2ConnectionFactory) connectionFactory);
        } else if (connectionFactory instanceof OAuth1ConnectionFactory) {
            return new OAuth1Flow(true, (OAuth1ConnectionFactory<?>) connectionFactory);
        } else {
            throw new AuthenticationException("failed to select operation");
        }
    }

    private void storeProviderName(HttpServletRequest request, String providerName) {
        request.getSession(true).setAttribute(OAUTH_PARAM_PROVIDER, providerName);
    }

    private String retrieveProviderName(HttpServletRequest request) {
        HttpSession session = request.getSession(false);
        if (session == null) {
            return null;
        }
        try {
            return (String) session.getAttribute(OAUTH_PARAM_PROVIDER);
        } finally {
            session.removeAttribute(OAUTH_PARAM_PROVIDER);
        }
    }

    private String callbackUrl(HttpServletRequest request) {
        URIBuilder base = URIBuilder.fromUri(callbackUrlBase(request));
        if (returnToUrlParameters != null) {
            for (String paramName : returnToUrlParameters) {
                if (request.getParameterMap().containsKey(paramName)) {
                    base.queryParam(paramName, request.getParameter(paramName));
                }
            }
        }
        return base.build().toString();
    }

    private String callbackUrlBase(HttpServletRequest request) {
        if (StringUtils.hasText(overrideSecureUrl)) {
            return overrideSecureUrl + FILTER_ROUTE;
        }
        return request.getRequestURL().toString();
    }
}

File: src/waypalm-site/src/main/java/waypalm/site/web/mvc/oauth/OAuthAuthenticationFilter.kt
package waypalm.site.web.mvc.oauth

import org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter
import org.springframework.social.connect.ConnectionFactoryLocator
import javax.inject.Inject
import org.springframework.security.web.authentication.rememberme.AbstractRememberMeServices
import java.util.HashSet
import org.springframework.social.connect.Connection
import javax.servlet.http.HttpServletRequest
import org.springframework.social.connect.support.OAuth1ConnectionFactory
import org.springframework.util.StringUtils
import org.springframework.social.connect.support.OAuth2ConnectionFactory
import org.springframework.social.support.URIBuilder
import javax.security.sasl.AuthenticationException
import javax.servlet.http.HttpServletResponse
import org.springframework.security.core.Authentication

public class OAuthAuthenticationFilter [Inject](
        var connectionFactoryLocator: ConnectionFactoryLocator
)
: AbstractAuthenticationProcessingFilter(OAuthAuthenticationFilter.FILTER_ROUTE)
{
    class object {
        val FILTER_ROUTE = "/j_spring_oauth_security_check"
        val OAUTH_PARAM_PROVIDER = "oauth_provider"
    }

    val returnToUrlParameters = HashSet<String>();
    var overrideSecureUrl: String ? = null;

    public fun  setReturnToUrlParameters(returnToUrlParameters: Set<String>) {
        this.returnToUrlParameters.clear();
        this.returnToUrlParameters.addAll(returnToUrlParameters);
    }

    public override fun afterPropertiesSet() {
        super.afterPropertiesSet();
        if (returnToUrlParameters.isEmpty() && getRememberMeServices() is AbstractRememberMeServices) {
            var returnToUrlParameters = HashSet<String>();
            returnToUrlParameters.add((getRememberMeServices() as AbstractRememberMeServices).getParameter()!!);
            setReturnToUrlParameters(returnToUrlParameters)
        }
    }

    fun selectProviderName(request: HttpServletRequest): String {
        var name = request.getParameter(OAUTH_PARAM_PROVIDER);
        if (name == null) {
            val nameFromRequest = retrieveProviderName(request);
            if(nameFromRequest == null)
                throw IllegalStateException("OAuth provider is unknown");
            return nameFromRequest
        }
        return name!!
    }

    /**
     * OAuth Authentication flow: See http://dev.twitter.com/pages/auth
     */
    public override fun attemptAuthentication(requestParam: HttpServletRequest?, responseParam: HttpServletResponse?): Authentication? {
        var request = requestParam!!
        var response = responseParam!!

        val providerName = selectProviderName(request)
        var flow = selectFlow(providerName, request);
        if (flow.initiationRequired) {
            storeProviderName(request, providerName);
            flow.initiateAuthentication(request, response);
            return null;
        }

        var connection: Connection<*> = flow.attemptAuthentication(request)!!;
        return getAuthenticationManager()!!.authenticate(OAuthAuthenticationToken(connection = connection));
    }

    private fun selectFlow(provider: String, request: HttpServletRequest): OAuthFlow<*> {
        val connectionFactory = connectionFactoryLocator.getConnectionFactory(provider);

        if (StringUtils.hasText(request.getParameter("oauth_token"))) {
            return OAuth1Flow(false, connectionFactory as OAuth1ConnectionFactory<*>, callbackUrl(request));
        } else if (StringUtils.hasText(request.getParameter("code"))) {
            return OAuth2Flow(false, connectionFactory as OAuth2ConnectionFactory<*>, callbackUrl(request));
        } else if (connectionFactory is OAuth2ConnectionFactory<*>) {
            return OAuth2Flow(true, connectionFactory as OAuth2ConnectionFactory<*>, callbackUrl(request));
        } else if (connectionFactory is OAuth1ConnectionFactory<*>) {
            return OAuth1Flow(true, connectionFactory as OAuth1ConnectionFactory<*>, callbackUrl(request));
        } else {
            throw AuthenticationException("failed to select operation");
        }
    }

    private fun storeProviderName(request: HttpServletRequest, providerName: String) {
        request.getSession(true)!!.setAttribute(OAUTH_PARAM_PROVIDER, providerName);
    }

    private fun retrieveProviderName(request: HttpServletRequest): String? {
        val session = request.getSession(false);
        if (session == null) {
            return null;
        }
        try {
            return session.getAttribute(OAUTH_PARAM_PROVIDER) as String;
        } finally {
            session.removeAttribute(OAUTH_PARAM_PROVIDER);
        }
    }

    private fun callbackUrl(request: HttpServletRequest): String {
        val base = URIBuilder.fromUri(callbackUrlBase(request))!!;
        for (paramName  :String  in returnToUrlParameters) {
            if (request.getParameterMap()!!.containsKey(paramName)) {
                base.queryParam(paramName, request.getParameter(paramName));
            }
        }
        return base.build().toString();
    }

    private fun callbackUrlBase(request: HttpServletRequest): String {
        if (StringUtils.hasText(overrideSecureUrl)) {
            return overrideSecureUrl + FILTER_ROUTE;
        }
        return request.getRequestURL().toString();
    }
}